src/components/InputUpload/InputUpload.tsx

```diff
@@ -0,0 +1,109 @@
+"use client";
+
+import React, { useRef, useState } from "react";
+import {
+  FieldValues,
+  Path,
+  UseFormReturn,
+  UseFormRegisterReturn,
+  PathValue,
+} from "react-hook-form";
+import Image from "next/image";
+import { IMAGE_TYPES } from "./InputUpload.constants";
+import * as S from "./InputUpload.styles";
+interface InputUploadProps<T extends FieldValues> {
+  position?: "relative" | "absolute";
+  isPreview?: boolean;
+  register: UseFormRegisterReturn;
+  setValue: UseFormReturn<T>["setValue"];
+  name: Path<T>;
+  children?: React.ReactNode;
+  onPreview?: (preview: string | null, imageSize: { width: number; height: number }) => void;
+}
+
+const InputUpload = <T extends FieldValues>({
+  position = "relative",
+  isPreview = true,
+  register,
+  setValue,
+  name,
+  children,
+  onPreview,
+}: InputUploadProps<T>) => {
+  const inputRef = useRef<null | HTMLInputElement>(null);
+  const [preview, setPreview] = useState<string | null>(null);
+  const [imageSize, setImageSize] = useState({
+    width: 0,
+    height: 0,
+  });
+
+  const handleFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (event.target.files) {
+      const file = event.target.files[0];
+      const fileType = file.type;
+
+      if (IMAGE_TYPES[fileType]) {
+        const reader = new FileReader();
+        reader.onloadend = () => {
+          const image = document.createElement("img");
+          image.onload = () => {
+            setImageSize({
+              width: image.width,
+              height: image.height,
+            });
+
+            if (onPreview) {
+              onPreview(reader.result as string, { width: image.width, height: image.height });
+            } else {
+              setPreview(reader.result as string);
+            }
+          };
+          image.src = reader.result as string;
+        };
+        reader.readAsDataURL(file);
+      } else {
+        alert("파일 형식이 올바르지 않습니다. 이미지 파일을 업로드해 주세요.");
+        event.target.value = "";
+      }
+    }

```

Comment:
[ P3 ]

조건문을 얼리리턴으로 개선해볼 수 있지않을까 해서 한번 해보았습니다!

```suggestion
    if (!event.target.files) {
      return;
    }

    const file = event.target.files[0];
    const fileType = file.type;

    if (!IMAGE_TYPES[fileType]) {
      alert("파일 형식이 올바르지 않습니다. 이미지 파일을 업로드해 주세요.");
      event.target.value = "";
      return;
    }

    const reader = new FileReader();

    reader.onloadend = () => {
      const image = document.createElement("img");

      image.onload = () => {
        setImageSize({
          width: image.width,
          height: image.height,
        });

        if (onPreview) {
          onPreview(reader.result as string, { width: image.width, height: image.height });
        } else {
          setPreview(reader.result as string);
        }
      };

      image.src = reader.result as string;
    };

    reader.readAsDataURL(file);
```

Reply:
오오 감사합니다!!
조건문이 많긴 한데, 코드 읽기는 훨씬 좋은 것 같네요!!
패턴 사용하기로 했는데 놓치고 있었는데 감사합니다!! 👍🏻👍🏻